tools/clang/tools/hlsl2spirv/hlsl2spirv.cpp

```diff
@@ -0,0 +1,216 @@
+//===- hlsl2spirv.cpp ---- Entry point for hlsl2spirv program -------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "dxc/Support/Global.h"
+#include "dxc/Support/Unicode.h"
+#include "dxc/Support/WinIncludes.h"
+#include "dxc/Support/HLSLOptions.h"
+#include "dxc/Support/dxcapi.use.h"
+#include "dxc/dxcapi.h"
+#include "dxc/dxcapi.internal.h"
+
+#include "llvm/Option/OptTable.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "spirv-tools/libspirv.hpp"
+
+using namespace dxc;
+using namespace llvm;
+using namespace llvm::opt;
+using namespace hlsl::options;
+
+namespace {
+void ConvertIDxcBlobToUint32(const CComPtr<IDxcBlob> &blob,
+                             std::vector<uint32_t> *binaryWords) {
+  if (!blob)
+    return;
+  size_t num32BitWords = (blob->GetBufferSize() + 3) / 4;
+  std::string binaryStr((char *)blob->GetBufferPointer(),
+                        blob->GetBufferSize());
+  binaryStr.resize(num32BitWords * 4, 0);
+  binaryWords->resize(num32BitWords, 0);
+  memcpy(binaryWords->data(), binaryStr.data(), binaryStr.size());
+}
+} // anonymous namespace
+
+class Hlsl2SpirvContext {
+private:
+  DxcDllSupport &dllSupport;
+  DxcOpts &options;
+  CComPtr<IDxcBlob> binaryBlobPtr;
+  CComPtr<IDxcBlob> spirvAssemblyBlobPtr;
+  std::string spirvAssembly;
+
+public:
+  Hlsl2SpirvContext(DxcDllSupport &dxcSupport, DxcOpts &opts)
+      : dllSupport(dxcSupport), options(opts) {}
+
+  void Compile();
+  void DisassembleResults();
+  void ReportResults();
+};
+
+void Hlsl2SpirvContext::ReportResults() {
+  bool needDisassembly =
+      !options.AssemblyCode.empty() || options.OutputObject.empty();
+
+  if (needDisassembly)
+    DisassembleResults();
+
+  if (!options.AssemblyCode.empty())
+    WriteBlobToFile(spirvAssemblyBlobPtr, StringRefUtf16(options.AssemblyCode));
+
+  if (!options.OutputObject.empty())
+    WriteBlobToFile(binaryBlobPtr, StringRefUtf16(options.OutputObject));
+  else
+    WriteUtf8ToConsole(spirvAssembly.c_str(), spirvAssembly.size(),
+                       STD_OUTPUT_HANDLE);
+}
+
+void Hlsl2SpirvContext::DisassembleResults() {
+  // First, convert the compiled blob into a vector of uint32_t
+  std::vector<uint32_t> words;
+  ConvertIDxcBlobToUint32(binaryBlobPtr, &words);
+
+  spvtools::SpirvTools spirvTools(SPV_ENV_UNIVERSAL_1_0);
+  spirvTools.SetMessageConsumer(
+      [](spv_message_level_t, const char *, const spv_position_t &,
+         const char *message) { fprintf(stdout, "%s\n", message); });
+  uint32_t options = SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES;
+  spirvTools.Disassemble(words, &spirvAssembly, options);
+
+  // SPIR-V Tools returns the disassembly as a string. Convert to blob.
+  CComPtr<IDxcLibrary> library;
```

Comment:
I think this is only needed when we are writing the disassembly into a file? Can we make it a method and call when necessary in `Report`?